From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NickAc <32451103+NickAcPT@users.noreply.github.com>
Date: Fri, 19 Feb 2021 05:17:52 +0000
Subject: [PATCH] Add game profile Display Unique ID


diff --git a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
index 2751ce7f1556da07ef853807a588f096adf6ef7f..313202530c7c44eb1afd085913fbb79ace7e0856 100644
--- a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+++ b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
@@ -69,6 +69,19 @@ public class CraftPlayerProfile implements PlayerProfile {
         return prev.getId();
     }
 
+    @Override
+    public UUID getDisplayId() {
+        return profile.getDisplayId();
+    }
+
+    @Override
+    public UUID setDisplayId(UUID displayUuid) {
+        GameProfile prev = this.profile;
+        this.profile = new GameProfile(prev.getId(), displayUuid, prev.getName());
+        copyProfileProperties(prev, this.profile);
+        return prev.getId();
+    }
+
     @Nullable
     @Override
     public String getName() {
diff --git a/src/main/java/com/mojang/authlib/GameProfile.java b/src/main/java/com/mojang/authlib/GameProfile.java
index 1369f0e680bd581a7fb93fb37ef5ab3d6aee5c98..215fc1cec4c4eb5d3854d4004ee7395f1d0cc0c5 100644
--- a/src/main/java/com/mojang/authlib/GameProfile.java
+++ b/src/main/java/com/mojang/authlib/GameProfile.java
@@ -7,9 +7,10 @@ import org.apache.commons.lang3.builder.ToStringBuilder;
 import java.util.UUID;
 
 public class GameProfile {
+    private UUID displayId; // Arcadium
     private final UUID id;
     private final String name;
-    private final PropertyMap properties = new PropertyMap();
+    private PropertyMap properties = new PropertyMap(); // Arcadium - final -> not final
     private boolean legacy;
 
     /**
@@ -29,6 +30,20 @@ public class GameProfile {
         this.id = id;
         this.name = name;
     }
+    /**
+     * Constructs a new Game Profile with the specified ID and name.
+     * <p />
+     * Either ID or name may be null/empty, but at least one must be filled.
+     *
+     * @param id Unique ID of the profile
+     * @param displayId Display Unique ID of the profile
+     * @param name Display name of the profile
+     * @throws java.lang.IllegalArgumentException Both ID and name are either null or empty
+     */
+    public GameProfile(final UUID id, final UUID displayId, final String name) {
+        this(id, name);
+        this.displayId = displayId;
+    }
 
     /**
      * Gets the unique ID of this game profile.
@@ -41,6 +56,17 @@ public class GameProfile {
         return id;
     }
 
+    /**
+     * Gets the unique Display ID of this game profile.
+     * <p />
+     * This may be null for partial profile data if constructed manually.
+     *
+     * @return Display ID of the profile
+     */
+    public UUID getDisplayId() {
+        return displayId;
+    }
+
     /**
      * Gets the display name of this game profile.
      * <p />
@@ -113,4 +139,17 @@ public class GameProfile {
     public boolean isLegacy() {
         return legacy;
     }
+
+    /**
+     * Gets a copy of this profile if it includes a Display ID, or itself if it doesn't.
+     * @return A GameProfile with a different Unique ID if needed
+     */
+    public GameProfile getDisplayProfile() {
+        if (displayId != null) {
+            GameProfile copyProfile = new GameProfile(displayId, name);
+            copyProfile.properties = this.properties;
+            return copyProfile;
+        }
+        return this;
+    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index ebff0ef7f9bf38562265fc2cd4f81f59fda2278d..0e02e95dc023a29b3ebf7b2db66590f83c0067ef 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1913,6 +1913,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
     @Override
     public void sendMessage(IChatBaseComponent ichatbasecomponent, UUID uuid) {
+        if (uuid.equals(uniqueID)) uuid = getProfile().getDisplayProfile().getId(); // Arcadium - Use Display profile
         this.a(ichatbasecomponent, ChatMessageType.SYSTEM, uuid);
     }
 
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
index 9a48a4e90c1b1cbcc8dc604fdf0f20a6d0f3bd22..d6bb78d54590e83ca4da7d7c17b2555ec0f8cd45 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -182,6 +182,11 @@ public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
         }
 
         public GameProfile a() {
+            // Arcadium start - OBFHELPER
+            return getGameProfile().getDisplayProfile(); // Arcadium - Display profile
+        }
+        public GameProfile getGameProfile() {
+            // Arcadium end
             return this.d;
         }
 
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 1471785da53a18edf31a20212f3cc6c19de92822..013fa9456afc5aa6b01ddc5fa3d5f64e2134426f 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1963,7 +1963,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             } else {
                 ChatMessage chatmessage = new ChatMessage("chat.type.text", new Object[]{this.player.getScoreboardDisplayName(), s});
 
-                this.minecraftServer.getPlayerList().sendMessage(chatmessage, ChatMessageType.CHAT, this.player.getUniqueID());
+                this.minecraftServer.getPlayerList().sendMessage(chatmessage, ChatMessageType.CHAT, this.player.getProfile().getDisplayProfile().getDisplayId()); // Arcadium - Use Display profile
             }
 
             // Spigot start - spam exclusions
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 7db99b6639a0afdd1f68539f6c6a6f48275e5b6b..1d1fb8eb9596504534ba335c471b8687501316b0 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1010,13 +1010,13 @@ public abstract class PlayerList {
         ScoreboardTeamBase scoreboardteambase = entityhuman.getScoreboardTeam();
 
         if (scoreboardteambase == null) {
-            this.sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, entityhuman.getUniqueID());
+            this.sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, entityhuman.getProfile().getDisplayProfile().getId()); // Arcadium - Use Display profile
         } else {
             for (int i = 0; i < this.players.size(); ++i) {
                 EntityPlayer entityplayer = (EntityPlayer) this.players.get(i);
 
                 if (entityplayer.getScoreboardTeam() != scoreboardteambase) {
-                    entityplayer.sendMessage(ichatbasecomponent, entityhuman.getUniqueID());
+                    entityplayer.sendMessage(ichatbasecomponent, entityhuman.getProfile().getDisplayProfile().getId());
                 }
             }
 
