From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NickAc <32451103+NickAcPT@users.noreply.github.com>
Date: Fri, 19 Feb 2021 05:17:52 +0000
Subject: [PATCH] Add game profile Display Unique ID


diff --git a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
index 2751ce7f1556da07ef853807a588f096adf6ef7f..c6f641d45296abafed4e24ca3ba3ba9e9311539e 100644
--- a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+++ b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
@@ -33,6 +33,10 @@ public class CraftPlayerProfile implements PlayerProfile {
         this.profile = new GameProfile(id, name);
     }
 
+    public CraftPlayerProfile(UUID id, UUID displayId, String name) {
+        this.profile = new GameProfile(id, displayId, name);
+    }
+
     public CraftPlayerProfile(GameProfile profile) {
         Validate.notNull(profile, "GameProfile cannot be null!");
         this.profile = profile;
@@ -69,6 +73,19 @@ public class CraftPlayerProfile implements PlayerProfile {
         return prev.getId();
     }
 
+    @Override
+    public UUID getDisplayId() {
+        return profile.getDisplayId();
+    }
+
+    @Override
+    public UUID setDisplayId(UUID displayUuid) {
+        GameProfile prev = this.profile;
+        this.profile = new GameProfile(prev.getId(), displayUuid, prev.getName());
+        copyProfileProperties(prev, this.profile);
+        return prev.getId();
+    }
+
     @Nullable
     @Override
     public String getName() {
@@ -124,7 +141,7 @@ public class CraftPlayerProfile implements PlayerProfile {
 
     @Override
     public CraftPlayerProfile clone() {
-        CraftPlayerProfile clone = new CraftPlayerProfile(this.getId(), this.getName());
+        CraftPlayerProfile clone = new CraftPlayerProfile(this.getId(), this.getDisplayId(), this.getName());
         clone.setProperties(getProperties());
         return clone;
     }
diff --git a/src/main/java/com/mojang/authlib/GameProfile.java b/src/main/java/com/mojang/authlib/GameProfile.java
index 1369f0e680bd581a7fb93fb37ef5ab3d6aee5c98..215fc1cec4c4eb5d3854d4004ee7395f1d0cc0c5 100644
--- a/src/main/java/com/mojang/authlib/GameProfile.java
+++ b/src/main/java/com/mojang/authlib/GameProfile.java
@@ -7,9 +7,10 @@ import org.apache.commons.lang3.builder.ToStringBuilder;
 import java.util.UUID;
 
 public class GameProfile {
+    private UUID displayId; // Arcadium
     private final UUID id;
     private final String name;
-    private final PropertyMap properties = new PropertyMap();
+    private PropertyMap properties = new PropertyMap(); // Arcadium - final -> not final
     private boolean legacy;
 
     /**
@@ -29,6 +30,20 @@ public class GameProfile {
         this.id = id;
         this.name = name;
     }
+    /**
+     * Constructs a new Game Profile with the specified ID and name.
+     * <p />
+     * Either ID or name may be null/empty, but at least one must be filled.
+     *
+     * @param id Unique ID of the profile
+     * @param displayId Display Unique ID of the profile
+     * @param name Display name of the profile
+     * @throws java.lang.IllegalArgumentException Both ID and name are either null or empty
+     */
+    public GameProfile(final UUID id, final UUID displayId, final String name) {
+        this(id, name);
+        this.displayId = displayId;
+    }
 
     /**
      * Gets the unique ID of this game profile.
@@ -41,6 +56,17 @@ public class GameProfile {
         return id;
     }
 
+    /**
+     * Gets the unique Display ID of this game profile.
+     * <p />
+     * This may be null for partial profile data if constructed manually.
+     *
+     * @return Display ID of the profile
+     */
+    public UUID getDisplayId() {
+        return displayId;
+    }
+
     /**
      * Gets the display name of this game profile.
      * <p />
@@ -113,4 +139,17 @@ public class GameProfile {
     public boolean isLegacy() {
         return legacy;
     }
+
+    /**
+     * Gets a copy of this profile if it includes a Display ID, or itself if it doesn't.
+     * @return A GameProfile with a different Unique ID if needed
+     */
+    public GameProfile getDisplayProfile() {
+        if (displayId != null) {
+            GameProfile copyProfile = new GameProfile(displayId, name);
+            copyProfile.properties = this.properties;
+            return copyProfile;
+        }
+        return this;
+    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index ebff0ef7f9bf38562265fc2cd4f81f59fda2278d..0e02e95dc023a29b3ebf7b2db66590f83c0067ef 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1913,6 +1913,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
 
     @Override
     public void sendMessage(IChatBaseComponent ichatbasecomponent, UUID uuid) {
+        if (uuid.equals(uniqueID)) uuid = getProfile().getDisplayProfile().getId(); // Arcadium - Use Display profile
         this.a(ichatbasecomponent, ChatMessageType.SYSTEM, uuid);
     }
 
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
index fc0197c46c39eb59e94ea3690391b742d96b3a59..19a4c702e4142ca1de7c9530378fb31234a80f04 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutNamedEntitySpawn.java
@@ -17,7 +17,7 @@ public class PacketPlayOutNamedEntitySpawn implements Packet<PacketListenerPlayO
 
     public PacketPlayOutNamedEntitySpawn(EntityHuman entityhuman) {
         this.a = entityhuman.getId();
-        this.b = entityhuman.getProfile().getId();
+        this.b = entityhuman.getProfile().getDisplayProfile().getId();
         this.c = entityhuman.locX();
         this.d = entityhuman.locY();
         this.e = entityhuman.locZ();
diff --git a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
index 9a48a4e90c1b1cbcc8dc604fdf0f20a6d0f3bd22..4858a95e7918f879d08f0f0db52fcfc919a52615 100644
--- a/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/src/main/java/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -11,8 +11,8 @@ import javax.annotation.Nullable;
 
 public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
 
-    private PacketPlayOutPlayerInfo.EnumPlayerInfoAction a;
-    private final List<PacketPlayOutPlayerInfo.PlayerInfoData> b = Lists.newArrayList();
+    private PacketPlayOutPlayerInfo.EnumPlayerInfoAction a; public PacketPlayOutPlayerInfo.EnumPlayerInfoAction getAction() { return a; } // Arcadium - OBFHelper
+    private final List<PacketPlayOutPlayerInfo.PlayerInfoData> b = Lists.newArrayList(); public List<PlayerInfoData> getPlayerInfoDataList() { return b; }// Arcadium - OBFHELPER
 
     public PacketPlayOutPlayerInfo() {}
 
@@ -25,6 +25,15 @@ public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
             EntityPlayer entityplayer = aentityplayer1[j];
 
             this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
+
+            if (a == EnumPlayerInfoAction.REMOVE_PLAYER && entityplayer.getProfile().getDisplayId() != null) {
+                //If there is a display id, remove the original uuid from the list of players
+                PlayerInfoData playerData = new PlayerInfoData(entityplayer.getProfile(), entityplayer.ping,
+                        entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName());
+                playerData.setForceRealProfile(true);
+                this.b.add(playerData);
+
+            }
         }
 
     }
@@ -171,9 +180,21 @@ public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
 
         private final int b;
         private final EnumGamemode c;
-        private final GameProfile d;
+        private GameProfile d; // Arcadium - final -> not final
         private final IChatBaseComponent e;
 
+        // Arcadium start
+        private boolean forceRealProfile = false;
+
+        public boolean isForceRealProfile() {
+            return forceRealProfile;
+        }
+
+        public void setForceRealProfile(boolean forceRealProfile) {
+            this.forceRealProfile = forceRealProfile;
+        }
+        // Arcadium end
+
         public PlayerInfoData(GameProfile gameprofile, int i, EnumGamemode enumgamemode, @Nullable IChatBaseComponent ichatbasecomponent) {
             this.d = gameprofile;
             this.b = i;
@@ -182,9 +203,18 @@ public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
         }
 
         public GameProfile a() {
+            // Arcadium start - OBFHELPER
+            return forceRealProfile ? getGameProfile() : getGameProfile().getDisplayProfile(); // Arcadium - Display profile and force real profile
+        }
+        public GameProfile getGameProfile() {
+            // Arcadium end
             return this.d;
         }
 
+        public void setGameProfile(GameProfile profile) {
+            this.d = profile;
+        }
+
         public int b() {
             return this.b;
         }
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 1471785da53a18edf31a20212f3cc6c19de92822..c29c8bb55673476a06759a0a46dcebd1fd55e693 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -1835,6 +1835,21 @@ public class PlayerConnection implements PacketListenerPlayIn {
             }
         }
 
+        // Arcadium start
+        if (packet instanceof PacketPlayOutPlayerInfo) {
+            PacketPlayOutPlayerInfo playerInfo = (PacketPlayOutPlayerInfo) packet;
+            List<PacketPlayOutPlayerInfo.PlayerInfoData> dataList = playerInfo.getPlayerInfoDataList();
+            if (dataList != null) {
+                dataList.forEach(info -> {
+                    // Do not change profile for self
+                    if (info.getGameProfile().getDisplayId() != null && player.getUniqueID() == info.getGameProfile().getId()) {
+                        info.setForceRealProfile(true);
+                    }
+                });
+            }
+        }
+        // Arcadium end
+
         // CraftBukkit start
         if (packet == null || this.processedDisconnect) { // Spigot
             return;
@@ -1963,7 +1978,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
             } else {
                 ChatMessage chatmessage = new ChatMessage("chat.type.text", new Object[]{this.player.getScoreboardDisplayName(), s});
 
-                this.minecraftServer.getPlayerList().sendMessage(chatmessage, ChatMessageType.CHAT, this.player.getUniqueID());
+                this.minecraftServer.getPlayerList().sendMessage(chatmessage, ChatMessageType.CHAT, this.player.getProfile().getDisplayProfile().getDisplayId()); // Arcadium - Use Display profile
             }
 
             // Spigot start - spam exclusions
diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 7db99b6639a0afdd1f68539f6c6a6f48275e5b6b..1d1fb8eb9596504534ba335c471b8687501316b0 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1010,13 +1010,13 @@ public abstract class PlayerList {
         ScoreboardTeamBase scoreboardteambase = entityhuman.getScoreboardTeam();
 
         if (scoreboardteambase == null) {
-            this.sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, entityhuman.getUniqueID());
+            this.sendMessage(ichatbasecomponent, ChatMessageType.SYSTEM, entityhuman.getProfile().getDisplayProfile().getId()); // Arcadium - Use Display profile
         } else {
             for (int i = 0; i < this.players.size(); ++i) {
                 EntityPlayer entityplayer = (EntityPlayer) this.players.get(i);
 
                 if (entityplayer.getScoreboardTeam() != scoreboardteambase) {
-                    entityplayer.sendMessage(ichatbasecomponent, entityhuman.getUniqueID());
+                    entityplayer.sendMessage(ichatbasecomponent, entityhuman.getProfile().getDisplayProfile().getId());
                 }
             }
 
