From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NickAcPT <32451103+NickAcPT@users.noreply.github.com>
Date: Wed, 5 May 2021 18:55:41 +0100
Subject: [PATCH] Disable some useless commands


diff --git a/src/main/java/net/minecraft/commands/CommandDispatcher.java b/src/main/java/net/minecraft/commands/CommandDispatcher.java
index 079b8444d13121f0f01860e88cea25cce9d0cc7e..7261e326dec620156060e0aa17d0698e0dc948cd 100644
--- a/src/main/java/net/minecraft/commands/CommandDispatcher.java
+++ b/src/main/java/net/minecraft/commands/CommandDispatcher.java
@@ -130,7 +130,7 @@ public class CommandDispatcher {
         CommandGamemodeDefault.a(this.b);
         CommandDifficulty.a(this.b);
         CommandEffect.a(this.b);
-        CommandMe.a(this.b);
+        //CommandMe.a(this.b); // Arcadium - Disable some commands
         CommandEnchant.a(this.b);
         CommandXp.a(this.b);
         CommandFill.a(this.b);
@@ -152,7 +152,7 @@ public class CommandDispatcher {
         CommandReload.a(this.b);
         CommandRecipe.a(this.b);
         CommandReplaceItem.a(this.b);
-        CommandSay.a(this.b);
+        //CommandSay.a(this.b); // Arcadium - Disable some commands
         CommandSchedule.a(this.b);
         CommandScoreboard.a(this.b);
         CommandSeed.a(this.b, commanddispatcher_servertype != CommandDispatcher.ServerType.INTEGRATED);
@@ -165,7 +165,7 @@ public class CommandDispatcher {
         CommandSummon.a(this.b);
         CommandTag.a(this.b);
         CommandTeam.a(this.b);
-        CommandTeamMsg.a(this.b);
+        //CommandTeamMsg.a(this.b); // Arcadium - Disable some commands
         CommandTeleport.a(this.b);
         CommandTellRaw.a(this.b);
         CommandTime.a(this.b);
@@ -190,7 +190,7 @@ public class CommandDispatcher {
             CommandSaveOn.a(this.b);
             CommandIdleTimeout.a(this.b);
             CommandStop.a(this.b);
-            CommandWhitelist.a(this.b);
+            //CommandWhitelist.a(this.b); // Arcadium - Disable some commands
             net.pl3x.purpur.command.CreditsCommand.register(getDispatcher()); // Purpur
             net.pl3x.purpur.command.DemoCommand.register(getDispatcher()); // Purpur
             net.pl3x.purpur.command.PingCommand.register(getDispatcher()); // Purpur
